### DISCUSSION TITLE
foundry-fund-me-cu/test/unit /FundMeTest.t.sol small error

### DISCUSSION THREAD

**Discussion by DCVglobalnetwork** [Cyfrin/foundry-full-course-cu#2310] [General]
Hi , I stumbled on small error at the begining of my FundMe foundry test 

![image](https://github.com/user-attachments/assets/5aaf0052-ba9a-42cc-9bc4-64c1ff2f374f)

I would appreciate it if someone could help me.

**Comment by DCVglobalnetwork** [comment]
HERE IS THE ERROR PLEASE IGNORE THE FIRST IMAGE 
![image](https://github.com/user-attachments/assets/387218d0-fb15-4852-be38-a56100ae9b51)

**Comment by DCVglobalnetwork** [comment]
I was able to solve the error, but would you give me feedback if I did it correctly? The test went through. thank you 

![image](https://github.com/user-attachments/assets/9d324c16-975d-4b13-864f-cc4c9b55f813)

**Reply by EngrPips** [reply]
This makes sense.

**Reply by DCVglobalnetwork** [reply]
thank you for your feedback. I really appreciate it.  I have one more question, as Patrick wrote the test as followed 
```sh 
function testOwnerIsMsgSender() public {
assertEq(fundMe.i_owner(), msg.sender); 
```
It was throwing an error for me and when I wnet to check in `FundMe.sol` we have as followed 
The constructor of the FundMe contract sets the i_owner to msg.sender:

```sh
constructor(address priceFeed) {
    s_priceFeed = AggregatorV3Interface(priceFeed);
    i_owner = msg.sender;
}
```
Modifier onlyOwner
```sh
modifier onlyOwner() {
    if (msg.sender != i_owner) revert FundMe__NotOwner();
    _;
}
```
But testing i_owner was throwing an error for me and I had to do as followed 
```sh
  function testOwnerIsMsgSender() public view {
        assertEq(fundMe.getOwner(), address(this));
    }
```
![image](https://github.com/user-attachments/assets/4d697bd5-4747-4ebf-bd76-ca7855ab7ece)

My question is, is there perhaps something wrong with code or something? why did it work for Patrick the way he tested but did not for me?  Please if you could give me your feedback, I would appreciate it. 

THANK YOU

**Reply by EngrPips** [reply]
Hey @DCVglobalnetwork, Can you check your `setUp` function and Patrick's `setUp` function to see if there is any difference?

**Reply by DCVglobalnetwork** [reply]
Yes, I forgot to add Patrick changed the test as follows:

```sh
function testOwnerIsMsgSender() public {
assertEq(fundMe.i_owner(), address(this));
```

but again, throw an error for me as he used i_owner() instead of get.Owner(); 
Do you know why it was throwing an error for me ? Thank you

**Reply by EngrPips** [reply]
What is the visibility of `i_owner()` in your `FundMe` contract?

**Reply by DCVglobalnetwork** [reply]
ok so here we have Patrick`s 

```sh
function setUp() external {
fundMe = new FundMe();
}
function testOwnerIsMsgSender() public {
assertEq(fundMe.i_owner(), address(this));
}
```

now mine setUp 

```sh 
function setUp() external {
        fundMe = new FundMe(mockPriceFeed);
    }

    function testOwnerIsMsgSender() public view {
        assertEq(fundMe.getOwner(), address(this));
    }
}
``` 
I had to add this line 
```sh
    fundMe = new FundMe(mockPriceFeed);
```
as it was throwing an error for me 
should i try without it ? But the other test will be highlighted in red ?  What do you suggest ?

**Reply by DCVglobalnetwork** [reply]
i_owner is a private variable in `FundMe.sol`

**Reply by EngrPips** [reply]
Does Patrick's `FundMe` contract not take a constructor argument?

**Reply by DCVglobalnetwork** [reply]
No, it does not, and perhaps why is that correct?

**Reply by EngrPips** [reply]
> i_owner is a private variable in `FundMe.sol`

If you make it public you will be able to do `fundMe.i_owner()` too without running into error.

**Reply by DCVglobalnetwork** [reply]
but later on he will add but when he told us to copy and paste from repo, it was already with an argument

**Reply by EngrPips** [reply]
> No, it does not, and perhaps why is that correct?

Maybe he has the priceFeed address as a constant variable in his contract, which is equally correct, but passing it as a constructor argument seems more appropriate.

**Reply by EngrPips** [reply]
> but later on he will add but when he told us to copy and paste from repo, it was already with an argument

I see, Keep following the course and things will become clear with time.

**Reply by DCVglobalnetwork** [reply]
Ok so I did change from private to public and the way he wrote the test went through but the contract from Foundry FundMe I copied from the get-go is with private, not public

**Reply by EngrPips** [reply]
The test from GitHub used `fundMe.i_owner()`?

**Reply by DCVglobalnetwork** [reply]
Okay, so what is happening here? He told us to copy and paste from repo and the contract has an argument and private i_owner but from the contract, he is using as follows 
```sh 
constructor() {
i_owner = msg.sender;
}
```
doesnt have argument so now i understand why it was throwing an error because if there is no argument, the way he wrote the test should be working for me but he did not say to do so , therefore i got a little confused

**Reply by DCVglobalnetwork** [reply]
No, the test does not have the contract, `FundMe.sol` 

the section Foundry Fundamentals, Section 2: Foundry Fund Me

the test from repo is complete 
In this course, he is writing the test from scratch, but it was throwing an error for me. I think if I try to do `FundMe` without an argument, it could pass the test. What do you suggest?

**Reply by EngrPips** [reply]
You can try it out and see.

**Reply by DCVglobalnetwork** [reply]
Hi , yes I did and it went through, but I stumbled on another small error with fork testing . I went to test `FundMeTest.t.sol` 

![image](https://github.com/user-attachments/assets/f7576cf5-f462-4daa-b836-d4f85825c0ff)

i used alchemy sepolia rpc URL, but it is throwing me the same error, which is reverting the function as there is no address but I am doing exactly as Patrick on the video. Could you please help? I would appreciate it.

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/user-attachments/assets/54215dbc-be63-4378-bb97-e1c1c39e8b53)

**Reply by DCVglobalnetwork** [reply]
here is `FundMe.sol` 
```sh 
// SPDX-License-Identifier: MIT
// 1. Pragma
pragma solidity 0.8.19;
// 2. Imports

import {AggregatorV3Interface} from "../lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

// 3. Interfaces, Libraries, Contracts
error FundMe__NotOwner();

/**
 * @title A sample Funding Contract
 * @author Patrick Collins
 * @notice This contract is for creating a sample funding contract
 * @dev This implements price feeds as our library
 */
contract FundMe {
    // Type Declarations
    using PriceConverter for uint256;

    // State variables
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    address public immutable i_owner;
    address[] private s_funders;
    mapping(address => uint256) private s_addressToAmountFunded;
    AggregatorV3Interface private s_priceFeed;

    // Events (we have none!)

    // Modifiers
    modifier onlyOwner() {
        // require(msg.sender == i_owner);
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }

    // Functions Order:
    //// constructor
    //// receive
    //// fallback
    //// external
    //// public
    //// internal
    //// private
    //// view / pure

    constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
        i_owner = msg.sender;
    }

    /// @notice Funds our contract based on the ETH/USD price
    function fund() public payable {
        require(
            msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD,
            "You need to spend more ETH!"
        );
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        s_addressToAmountFunded[msg.sender] += msg.value;
        s_funders.push(msg.sender);
    }

    function withdraw() public onlyOwner {
        for (
            uint256 funderIndex = 0;
            funderIndex < s_funders.length;
            funderIndex++
        ) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // Transfer vs call vs Send
        // payable(msg.sender).transfer(address(this).balance);
        (bool success, ) = i_owner.call{value: address(this).balance}("");
        require(success);
    }

    function cheaperWithdraw() public onlyOwner {
        address[] memory funders = s_funders;
        // mappings can't be in memory, sorry!
        for (
            uint256 funderIndex = 0;
            funderIndex < funders.length;
            funderIndex++
        ) {
            address funder = funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // payable(msg.sender).transfer(address(this).balance);
        (bool success, ) = i_owner.call{value: address(this).balance}("");
        require(success);
    }

    /**
     * Getter Functions
     */

    /**
     * @notice Gets the amount that an address has funded
     *  @param fundingAddress the address of the funder
     *  @return the amount funded
     */
    function getAddressToAmountFunded(
        address fundingAddress
    ) public view returns (uint256) {
        return s_addressToAmountFunded[fundingAddress];
    }

    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }

    function getFunder(uint256 index) public view returns (address) {
        return s_funders[index];
    }

    function getOwner() public view returns (address) {
        return i_owner;
    }

    function getPriceFeed() public view returns (AggregatorV3Interface) {
        return s_priceFeed;
    }
}
```

and here is `FundMeTest.t.sol`

```sh
// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {AggregatorV3Interface} from "../lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

contract FundMeTest is Test {
    FundMe fundMe;
    address mockPriceFeed;

    function setUp() external {
   
        fundMe = new FundMe(mockPriceFeed);
    }

    function testMiniumDollarIsFive() public view {
        assertEq(fundMe.MINIMUM_USD(), 5E18);
    }
    function testOwnerIsMsgSender() public view {
        assertEq(fundMe.i_owner(), address(this));
        // we should not checking if the msg.sender is the owner but the test is the owner that why address(this)
        // that why using console with test can help you identify the error
    }
    function testVersionIsCorrect() public view {
        uint256 expectedVersion = 4; // Replace this with the expected version of your PriceFeed
        uint256 actualVersion = fundMe.getVersion();
        assertEq(
            actualVersion,
            expectedVersion,
            "PriceFeed version is incorrect"
        );
    }
}
```

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/user-attachments/assets/538366b6-1486-47a8-92e2-687123908aeb)

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/user-attachments/assets/4d35deb6-d3e5-4985-ba06-9ae086bba48e)

![image](https://github.com/user-attachments/assets/b07a38ec-1938-4070-b1f4-5d24358ea778)

these two look different 
function getVersion() 

Following Patrick from Youtube, it is throwing me an error when I am testing getVersion and also using fork testing, but on Alchemy, it seems to be working. I am a bit confused I will start over again with `FundMe` but please, if you have any suggestion , yes, I did aks Ai but is giving me the end result with mocks but from the video, im not there yet with mocks . Thank you

**Reply by EngrPips** [reply]
Has Patrick done a `fork test` yet in the course?

**Reply by DCVglobalnetwork** [reply]
Okay, so what happened here was `FundMe.sol` from Remix not `FundMe.sol` from Foundry . I am going over again with this part of the lesson, and I will see now that `FundMe.sol` from Remix should work.

**Reply by DCVglobalnetwork** [reply]
ok so here is the full code 
```sh 
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// Note: The AggregatorV3Interface might be at a different location than what was in the video!
import {AggregatorV3Interface} from "../lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) public addressToAmountFunded;
    address[] public funders;

    // Could we make this constant?  /* hint: no! We should make it immutable! */
    address public /* immutable */ i_owner;
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;

    constructor() {
        i_owner = msg.sender;
    }

    function fund() public payable {
        require(msg.value.getConversionRate() >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }

    function getVersion() public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        return priceFeed.version();
    }

    modifier onlyOwner() {
        // require(msg.sender == owner);
        if (msg.sender != i_owner) revert NotOwner();
        _;
    }

    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);

        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
    // Explainer from: https://solidity-by-example.org/fallback/
    // Ether is sent to contract
    //      is msg.data empty?
    //          /   \
    //         yes  no
    //         /     \
    //    receive()?  fallback()
    //     /   \
    //   yes   no
    //  /        \
    //receive()  fallback()

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }
}
```
i have an error with this line 

```sh
require(msg.value.getConversionRate() >= MINIMUM_USD, "You need to spend more ETH!");
```
![image](https://github.com/user-attachments/assets/00bb5065-e5f5-4e5f-80b2-9a690a62cd6b)

**Reply by EngrPips** [reply]
Doesn't this `getConversionRate` take two arguments? If you look at your `PriceConverter` Library, you will see that it takes two arguments, but you are only passing one argument.

**Reply by DCVglobalnetwork** [reply]
Please kindly ignore the message above I was able to sort out

**Reply by EngrPips** [reply]
Alright, Keep crushing it.

**Reply by DCVglobalnetwork** [reply]
hello back , i am having an error with this line 
```sh
/**
     * Getter Functions
     */

    /**
     * @notice Gets the amount that an address has funded
     *  @param fundingAddress the address of the funder
     *  @return the amount funded
     */
    function getAddressToAmountFunded(address fundingAddress) external view returns (uint256) {
        return s_addressToAmountFunded[fundingAddress];
    }
```
here is full code `FundMe.sol`
```sh
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// Note: The AggregatorV3Interface might be at a different location than what was in the video!
import {AggregatorV3Interface} from "../lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) private s_addressToAmountFunded;
    address[] public s_funders;

    // Could we make this constant? * hint: no! We should make it immutable!
    address public immutable i_owner;
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    AggregatorV3Interface private s_priceFeed;

    constructor(address priceFeed) {
        i_owner = msg.sender;
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }

    function fund() public payable {
        require(
            msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD,
            "You need to spend more ETH!"
        );
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        s_addressToAmountFunded[msg.sender] += msg.value;
        s_funders.push(msg.sender);
    }

    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }

    modifier onlyOwner() {
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }

    function withdraw() public onlyOwner {
        for (
            uint256 funderIndex = 0;
            funderIndex < s_funders.length
            funderIndex++
        ) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);

        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess) = payable(msg.sender).call{
            value: address(this).balance
        }("");
        require(callSuccess, "Call failed");
    }

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }
}
/**
     * Getter Functions
     */

    /**
     * @notice Gets the amount that an address has funded
     *  @param fundingAddress the address of the funder
     *  @return the amount funded
     */
    function getAddressToAmountFunded(address fundingAddress) external view returns (uint256) {
        return s_addressToAmountFunded[fundingAddress];
    }
```

also before this line 
```sh
  mapping(address => uint256) public s_addressToAmountFunded;
    address[] public s_funders;
```
has been changed to this from `public to private ` as gas efficiently 
```sh
  mapping(address => uint256) private s_addressToAmountFunded;
    address[] public s_funders;
```
and i am getting an error with visibility 

![image](https://github.com/user-attachments/assets/181dc49f-219e-41b7-8e47-01e270784d5c)

could you assist me with the solution?

**Reply by EngrPips** [reply]
Hey @DCVglobalnetwork, What does the error say?

**Reply by Shiva-Sai-ssb** [reply]
Hey @DCVglobalnetwork, you are getting the "free functions cannot have visibility" error because the `getAddressToAmountFunded` function is defined outside the contract. In Solidity, functions must be declared within a contract or library to use visibility modifiers such as `external`, `public`, `internal`, or `private`.

To resolve this issue, move the `getAddressToAmountFunded` function inside the `FundMe` contract. This should fix the problem.

**Reply by DCVglobalnetwork** [reply]
no did not fix anything, still the same error

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/user-attachments/assets/2a917bbd-4769-4963-878a-2422e1913eff)

**Comment by EngrPips** [comment]
Interesting.

**Comment by DCVglobalnetwork** [comment]
![image](https://github.com/user-attachments/assets/3de53433-d856-42fb-ab82-61abfaeb4760)

**Comment by DCVglobalnetwork** [comment]
ok wow, I was able to solve the error yes by moving the getAddressToAmountFunded function inside the `FundMe` contract and not only as I got another error on line with 
```sh
s_founders 
```
where I should use array in order to make it work, and than I found another similar contract and went to final `FundMe.sol` in repo and it was the same as I found the error so after all that search and combination, finaly has been solved 

thank you both of you for support

**Comment by DCVglobalnetwork** [comment]
![image](https://github.com/user-attachments/assets/84510d96-ca61-4cbb-9e10-8410b9294bdd)

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/user-attachments/assets/bfc14050-0d6e-4e04-8257-e37ad5e0cbda)

**Comment by DCVglobalnetwork** [comment]
while writing another test, I stumble on another error 

![image](https://github.com/user-attachments/assets/6003570d-a9c8-475b-a5fa-16a356e2baba)

can anyone assist me with solution? 

testing 
```sh
function testAddsFunderToArrayOfFunders() public {
        fundMe.fund{value: SEND_VALUE}();

        address funder = fundMe.getFunder(0);
        assertEq(funder, USER);
    }
```
![image](https://github.com/user-attachments/assets/ff073c0e-ad4f-4f71-80c2-4cdda6f64c75)

**Reply by EngrPips** [reply]
Hello @DCVglobalnetwork, This is an assertion error, so you should be able to figure this out yourself. Figure out why your assertion is failing.

**Reply by DCVglobalnetwork** [reply]
yes I did it , Thank you 

![image](https://github.com/user-attachments/assets/7b8982bc-a8cf-45fe-a893-a2c7260172da)

**Reply by EngrPips** [reply]
You are really capable of this things if only you push your limit. Keep crushing it!

**Reply by DCVglobalnetwork** [reply]
Thank you for your courage

**Reply by EngrPips** [reply]
You are always welcome. Keep crushing it
